plugins/gnupgplugin/gpgme_p.h

```diff
@@ -0,0 +1,277 @@
+            gpgme_engine_info_t info;
+            gpgme_error_t err2;
+            err2 = gpgme_get_engine_info(&info);
+            if (!err2) {
+                while (info) {
+                    if (info->protocol == protocol) {
+                        qWarning() << "protocol:" << gpgme_get_protocol_name(info->protocol);
+                        if (info->file_name && !info->version)
+                            qWarning() << "engine" << info->file_name
+                                       << "not installed properly";
+                        else if (info->file_name && info->version && info->req_version)
+                            qWarning() << "engine" << info->file_name
+                                       << "version" << info->version
+                                       << "installed, but at least version"
+                                       << info->req_version << "required";
+                        else
+                            qWarning() << "unknow issue";
+                    }
+                    info = info->next;
+                }
+            } else {
+                qWarning() << "cannot get engine info:" << gpgme_strerror(err2);
+            }
+            return;
+        }
+
+        err = gpgme_new(&ctx);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            ctx = 0;
+            return;
+        }
+        err = gpgme_set_protocol(ctx, protocol);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            gpgme_release(ctx);
+            ctx = 0;
+            return;
+        }
+        if (home.isEmpty())
+            return;
+        gpgme_engine_info_t info = gpgme_ctx_get_engine_info(ctx);
+        while (info && info->protocol != protocol)
+            info = info->next;
+        if (!info) {
+            qWarning() << "cannot change home.";
+            return;
+        }
+        err = gpgme_ctx_set_engine_info(ctx, protocol, info->file_name,
+                                        home.toUtf8().constData());
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR) {
+            gpgme_release(ctx);
+            ctx = 0;
+            return;
+        }
+    }
+    ~GPGmeContext()
+    {
+        if (ctx)
+            gpgme_release(ctx);
+    }
+    bool operator!()
+    {
+        return (ctx == 0);
+    }
+    operator gpgme_ctx_t() const
+    {
+        return ctx;
+    }
+    QString error()
+    {
+        return (err && gpgme_err_code(err) != GPG_ERR_NO_ERROR)
+            ? QString(gpgme_strerror(err)) : QString();
+    }
+};
+
+struct GPGmeData {
+    gpgme_data_t data;
+    gpgme_error_t err;
+    GPGmeData()
+        : data(0), err(0)
+    {
+        err = gpgme_data_new(&data);
+        if (gpgme_err_code(err) != GPG_ERR_NO_ERROR)
+            data = 0;
+    }
+    GPGmeData(const QByteArray &origin)
+        : data(0), err(0)
+    {
+        err = gpgme_data_new_from_mem(&data, origin.constData(), origin.length(), 0);
```

Comment:
In fact, no copy is done, see the 0 argument on last position. Doc in https://www.gnupg.org/documentation/manuals/gpgme/Memory-Based-Data-Buffers.html

I may #define NO_COPY 0 to be clear...

Reply:
Then if it's not copy, then you need to carry QByteArray around too, to make sure that the actual data don't get deallocated before GPGmeData.

Reply:
These are just helpers in gpgme_p.h to automatically have deallocation. I kept the same restrictions than in gpgme API. Callers are responsible to keep the data alive indeed. Since these helpers are not passed between routines I think it's fine.

Reply:
I've added a comment to relmind that origin QByteArray object should be alive for the whole life of the created structure and replaced the last 0 argument with a NO_COPY #define.

Thank you for your comments.